processor/src/stack/overflow.rs

```diff
@@ -1,211 +1,282 @@
 use alloc::{collections::BTreeMap, vec::Vec};
+use core::ops::RangeInclusive;
 
-use super::{Felt, FieldElement, ZERO};
+use miden_air::RowIndex;
 
-// OVERFLOW TABLE
+use super::{Felt, ZERO};
+use crate::ContextId;
+
+// overflow stack
 // ================================================================================================
 
-/// Stores stack values beyond the top 16 elements as well as the data which can be used in
-/// construction of the auxiliary trace column describing the state of the overflow table at every
-/// VM cycle.
+/// An element of the overflow stack.
+///
+/// Store the value pushed on the overflow stack, and the clock cycle at which it was pushed.
+#[derive(Debug, Clone)]
+struct StackEntry {
+    pub value: Felt,
+    pub clk: RowIndex,
+}
+
+impl StackEntry {
+    pub fn new(value: Felt, clk: RowIndex) -> Self {
+        Self { value, clk }
+    }
+
+    pub fn value(&self) -> Felt {
+        self.value
+    }
+}
+
+/// An overflow stack which stores the values of the stack elements that overflow the top 16
+/// elements of the stack per context.
 ///
-/// When `trace_enabled` is set to true, we also record all changes to the table so that we can
-/// reconstruct the overflow table at any clock cycle. This can be used for debugging purposes.
+/// This overflow stack does not keep track of the clock cycles at which the values were added or
+/// removed from the table; it is only concerned with the state at the latest clock cycle.
+///
+/// The overflow stack keeps track of the current clock cycle, and hence `advance_clock()` must be
+/// called whenever the clock cycle is incremented globally.
 #[derive(Debug)]
-pub struct OverflowTable {
-    /// A list of all rows that were added to and then removed from the overflow table.
-    all_rows: Vec<OverflowTableRow>,
-    /// A list of indices into the `all_rows` vector which describes the rows currently in the
-    /// overflow table.
-    active_rows: Vec<usize>,
-    /// A map which records the full state of the overflow table at every cycle during which an
-    /// update happened. This map is populated only when `trace_enabled` = true.
-    trace: BTreeMap<u64, Vec<Felt>>,
-    /// A flag which specifies whether we should record the full state of the overflow table
-    /// whenever an update happens. This is set to true only when executing programs for debug
-    /// purposes.
-    trace_enabled: bool,
-    /// Holds the address (the clock cycle) of the row at to top of the overflow table. When
-    /// entering new execution context, this value is set to ZERO, and thus, will differ from the
-    /// row address actually at the top of the table.
-    last_row_addr: Felt,
+pub struct OverflowStack {
+    overflow: BTreeMap<ContextId, Vec<StackEntry>>,
```

Comment:
How does a `BTreeMap` work with syscalls? When we make a syscall, we'd insert an entry into here with `ContextId=0`, right? But if we also an overflow entries for the initial root context, wouldn't they get overwritten?

If so, a solution could be to change this to `Vec<(ContextId, Vec<StackEntry>)` or maybe even just `Vec<Vec<StackEntry>>`. This should work because contexts work like a stack - i.e.,

- When entering a new context an empty `Vec<StackEntry>` would be pushed onto the `overflow` vector.
- When existing a context, the top `Vec<StackEntry>` will be discarded (this top stack entry should be empty at the time of exit, or this will be an error).

Reply:
Good point; I added 2 tests to cover this use case, and switched to effectively `BTreeMap<ContextId, Vec<Vec<StackEntry>>>`; except that to make this more readable, I renamed `OverflowStack` to `OverflowTable`, and now `OverflowStack = Vec<StackEntry>`.

Also as discussed offline, I believe I had the "off by 1" wrong for `start/restore_context`: I now consider a `CALL` or associated `END` operation as being part of the new context (as I think it's the correct interpretation to have). This removed the need for the distinction between how `start/restore_context()` and `push/pop()` write to history.